### ISSUE TITLE
Krazo does not work with OpenLiberty 21.* beta 

### ISSUE THREAD

**Issue by cgawron** [eclipse-ee4j/krazo#282]
When I try to use krazo with OpenLiberty, I get an similar error as [OpenLiberty issue #5505](https://github.com/OpenLiberty/open-liberty/issues/5505).

This might be an OpenLiberty issue, but as the README states that krazo works with OpenLiberty, I'm reporting this here.

I'm using these dependencies in my pom.xml:
```
        <dependency>
            <groupId>jakarta.mvc</groupId>
            <artifactId>jakarta.mvc-api</artifactId>
            <version>2.0.0</version>
        </dependency>
        <dependency>
            <groupId>org.eclipse.krazo</groupId>
            <artifactId>krazo-core</artifactId>
            <version>2.0.1</version>
        </dependency>
        <dependency>
            <groupId>org.eclipse.krazo</groupId>
            <artifactId>krazo-resteasy</artifactId>
            <version>2.0.1</version>
        </dependency>
```

Here is the exception that I get when I try to access my MVC application:
```
[INFO] [FEHLER  ] SRVE0777E: Es wurde eine Ausnahme von der Anwendungsklasse 'org.jboss.resteasy.core.ExceptionHandler.handleException:381' ausgelöst.
[INFO] org.jboss.resteasy.spi.UnhandledException: java.lang.RuntimeException: SRV.8.2: RequestWrapper objects must extend ServletRequestWrapper or HttpServletRequestWrapper
[INFO]  at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:381)
[INFO]  at [internal classes]
[INFO] Caused by: java.lang.RuntimeException: SRV.8.2: RequestWrapper objects must extend ServletRequestWrapper or HttpServletRequestWrapper
[INFO]  at com.ibm.wsspi.webcontainer.util.ServletUtil.unwrapRequest(ServletUtil.java:89)
[INFO]  at [internal classes]
[INFO]  at org.eclipse.krazo.engine.ServletViewEngine.forwardRequest(ServletViewEngine.java:121)
[INFO]  at org.eclipse.krazo.engine.JspViewEngine.processView(JspViewEngine.java:58)
[INFO]  at org.eclipse.krazo.core.ViewableWriter.writeTo(ViewableWriter.java:159)
[INFO]  at org.eclipse.krazo.core.ViewableWriter.writeTo(ViewableWriter.java:71)
[INFO]  at org.jboss.resteasy.core.interception.jaxrs.ServerWriterInterceptorContext.lambda$writeTo$1(ServerWriterInterceptorContext.java:79)
[INFO]  ... 1 more
```

**Comment by chkal** [comment]
Thanks for reporting.

This is some kind of "known issue" and also the reason why the MVC TCK doesn't pass on OpenLiberty. Unfortunately I didn't find time yet to dig deeper into this.

But I'm pretty sure that this isn't caused by Krazo, because we actually use the standard wrapper classes for the forward:

https://github.com/eclipse-ee4j/krazo/blob/62481fa0c32215da4f77ab2ee6e96e2ea580e26c/core/src/main/java/org/eclipse/krazo/engine/ServletViewEngine.java#L114-L122

**Comment by cgawron** [comment]
@chkal Thanks for the quick response, I opened an [issue for OpenLiberty](https://github.com/OpenLiberty/open-liberty/issues/19271).
Is there a version of OpenLiberty which is known to work with krazo 2.* ? 
I'm planning to show Jakarta MVC in a lecture on Jakarta EE, and I would prefer using OpenLiberty as I use it for the other topics.

**Comment by erdlet** [comment]
Hi @cgawron,

unfortunately as far as I know there is no version of Krazo 2.0.* which works well with OpenLiberty, because the [migration to RESTEasy](https://github.com/OpenLiberty/open-liberty/issues/11803) and [Jakarta EE 9](https://github.com/OpenLiberty/open-liberty/issues/9876) on OpenLiberty is not finished yet and we had no chance to get reliable test results so far. I hope this will change during the next weeks.

For demos of MVC / Krazo on OpenLiberty you could use the stable OpenLiberty with the `javax` namespace and Krazo CXF 1.0.*. This works nearly stable, only the Bean Validation is not working as intended (caused by CXF). Also, we didn't add any major features between the versions, so Krazo behaves identically between version 1.0.* and 2.0.*.

**Comment by chkal** [comment]
@erdlet Agreed!

It would really be cool to get a green TCK on Liberty. It really looks like some kind of Liberty bug which is currently causing all the failures. I hope to find more time in the next days/weeks to have a deeper look at the root cause.

**Comment by jesse-gallagher** [comment]
I ran into this as well, and I spent time digging into it yesterday. I [blogged about](https://frostillic.us/blog/posts/2021/11/30/journeys-debugging-open-liberty-and-mvc) my whole path, but the crux of it is:

Liberty's [`WebAppRequestDispatcher` class](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/webapp/WebAppRequestDispatcher.java#L1133) expects to be able to get an instance of their `IExtendedRequest` type from the passed-in request object. Presumably, CXF handled this differently, but RESTEasy passes in a Proxy object that can't be unwrapped down to the special WLP class - that's the immediate trouble.

I tried using `CDI.current().select(HttpServletRequest.class).get()` in Krazo's `ServletViewEngine` instead, which returns a different WLP request class, but that one doesn't implement `IExtendedRequest`.

Liberty _does_ have [an internal class](https://github.com/OpenLiberty/open-liberty/blob/9a5c27a8ebeb6aa7e63170e9567a79ab05922e5e/dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WebContainerRequestState.java#L186) that provides access to the current thread's `IExtendedRequest`, and I used that in Krazo's `ServletViewEngine` reflectively:

```java
try {
  Class<?> requestStateClass = Class.forName("com.ibm.wsspi.webcontainer.WebContainerRequestState", false, request.getClass().getClassLoader());
  Method getInstance = requestStateClass.getDeclaredMethod("getInstance", boolean.class);
  Object requestState = getInstance.invoke(null, false);
  Method getCurrentThreadsIExtendedRequest = requestStateClass.getDeclaredMethod("getCurrentThreadsIExtendedRequest");
  request = (HttpServletRequest)getCurrentThreadsIExtendedRequest.invoke(requestState);
} catch (Throwable e1) {
  // Not on WAS
}
rd.forward(new HttpServletRequestWrapper(request), new HttpServletResponseWrapper(response));
```

This is... obviously not a solution that should go into `krazo-core`, but it does technically solve the problem. I'm not sure what the best route should be here - I could see either a WLP-specific extension for Krazo or a change in WLP's behavior for finding the needed request object. Since they already have a method that can get to it on their side, perhaps they could change their implementation to fall back to that when their initial attempt to get it from the passed-in wrapper fails.

I've created a small project to reproduce the problem here: [https://github.com/jesse-gallagher/wlp-krazo-problem](https://github.com/jesse-gallagher/wlp-krazo-problem)

**Comment by erdlet** [comment]
Thanks a lot for digging into this! 

Probably we can implement some provider for the correct request object in `krazo-resteasy` and call its interface in `krazo-core`. IIRC we have / had something similar for CXF🤔

**Comment by chkal** [comment]
@jesse-gallagher Thanks again for sharing your analysis. This is really valuable.

I wonder if it really makes sense to discuss how to work around this issue in Krazo. Isn't this a clear spec compliance issue in Liberty? It is actually a common pattern to pass a custom request/response down the filter chain from a servlet filter. The servlet spec only requires to pass down subclasses of `HttpServletRequestWrapper` or `HttpServletResponseWrapper`. And that's what we are doing here. So maybe this should simply be reported as an issue to the Liberty team? I just a bit surprised, that the Servlet TCK doesn't seem to verify this behavior.

**Comment by jesse-gallagher** [comment]
From my perspective, that's about what I would think: Krazo is doing things by the book, and it's just an interaction of the two pieces provided by Liberty that breaks it, not Krazo. If there's a clean way in RESTEasy to get the original request that Krazo could do in the -resteasy extension, maybe it'd still make sense to do that, but otherwise it seems like a Liberty issue.

**Comment by erdlet** [comment]
> I wonder if it really makes sense to discuss how to work around this issue in Krazo. Isn't this a clear spec compliance issue in Liberty?

I'm completely your opinion regarding the compliance issue, but I think we are faster implementing a workaround which can be removed later. My experience with OL is, like with CXF, that their implementations work different than in any other server because of IBM optimizations.

So my proposal is to find a clean solution to provide the correct Request / Response Wrappers in Krazo until we have feedback from the OL team regarding the spec compliance.

**Comment by chkal** [comment]
> So my proposal is to find a clean solution to provide the correct Request / Response Wrappers in Krazo until we have feedback from the OL team regarding the spec compliance.

Agreed. My experience with the OL team is that they are sometimes quite fast with providing fixes. Maybe providing a good reproducer will help. And I guess that an issue report like "this plain Jakarta EE use case fails on OL" gets higher prio than "Krazo doesn't work on OL".

Possibly I'll find some time over the weekend to experiment with creating such a reproducer. But I cannot promise that.

**Comment by erdlet** [comment]
> And I guess that an issue report like "this plain Jakarta EE use case fails on OL" gets higher prio than "Krazo doesn't work on OL".

Oh yes, thats a good point.

I hope to find some time next week to implement the workaround. Even if the OL team responds fast, we can check faster if other things are broken too or the rest of the RESTEasy module works fine on OL.

**Comment by jesse-gallagher** [comment]
I was able to boil this down to a pair of examples: one JEE 9 and one JEE 8, with the former triggering the exception and the latter not. They don't use Krazo as such, but they do follow a similar path as Krazo for how the `HttpServletRequest` and `*Response` objects are ferried around:

[https://github.com/jesse-gallagher/wlp-resteasy-servletrequest](https://github.com/jesse-gallagher/wlp-resteasy-servletrequest)

While looking into this, I also came up with "one and a half" workarounds.

The first is a `HttpCommunicationUnwrapper` extension for Krazo, which attempts to identify when running on Liberty and use the above reflective mechanism to find the `IExtendedRequest` object. This works in my app, but I'm a _little_ wary of it because it's possible that the request eventually sent down via JAX-RS actually comes from a different `IExtendedRequest` - not very likely, I imagine, but possible.

The "half" workaround is something that I was able to get working in my example project but not with Krazo itself. In that case, I switched uses of `@Context HttpServletRequest req`, etc. to `@Inject RequestStashingBean` (equivalent to `JaxRsContextProducer` in Krazo), and then used getters on that class directly. I'm not sure why the equivalent didn't work in my local Krazo-based app, though, and I hope to investigate that further tomorrow. If that route could be made to work, that may be preferable overall, since it'd still be using non-implementation-specific mechanisms. Unless there's a specific additional gain from using `@Context HttpServletRequest`, that is.

**Comment by erdlet** [comment]
> The first is a HttpCommunicationUnwrapper extension for Krazo, which attempts to identify when running on Liberty and use the above reflective mechanism to find the IExtendedRequest object.

@jesse-gallagher thanks for trying that. This was the mechanism I talked about above. We used this before in the CXF extension because the request / response implementations of TomEE and OL worked different too. So maybe you could try to integrate this `HttpCommunicationUnwrapper` to `krazo-resteasy` and run the `krazo-testsuite` against OL and WildFly. It would be fantastic when this really works.

**Comment by jesse-gallagher** [comment]
Can do. So far, it's working with WildFly and GlassFish, but there's a CSRF-related test failure on Liberty. I don't know that it's related to my change, but I may as well look into it while I'm in here.

**Comment by erdlet** [comment]
Thanks for testing! If the CSRF issue you mentioned in the new issue is the only one left, this fix helps a lot.

**Comment by chkal** [comment]
> The "half" workaround is something that I was able to get working in my example project but not with Krazo itself. In that case, I switched uses of `@Context HttpServletRequest req`, etc. to `@Inject RequestStashingBean` (equivalent to `JaxRsContextProducer` in Krazo), and then used getters on that class directly. I'm not sure why the equivalent didn't work in my local Krazo-based app, though, and I hope to investigate that further tomorrow. If that route could be made to work, that may be preferable overall, since it'd still be using non-implementation-specific mechanisms. Unless there's a specific additional gain from using `@Context HttpServletRequest`, that is.

Injecting `HttpServletRequest` either via `@Context` (which will be deprecated soon) or `@Inject` can be problematic, because each Servlet filter can actually "replace" the request before passing control to the next filter and eventually to the corresponding servlet processing the request. This is something Krazo does as well. And because of this ability to "replace" the request, the semantics of `@Inject HttpServletRequest request` isn't entirely well-defined. Is it the "original" request passed to the first filter (which would somehow "break" the replacing of the request from a filter) or is it the request returned from the last filter in the chain (which would be weird as well, because what about using `@Inject HttpServletRequest` in a filter itself.

To sum up: Accessing the `HttpServletRequest` via injection could bring all sorts of problems and AFAIK currently using `@Context` is the safest way to go.

**Comment by erdlet** [comment]
Since we have this issue again reported in #364, I'll close this for now. We have the workarounds but in the end this is an OL issue.
